### PR TITLE
[MISC] End-to-End Tests Refactoring

### PR DESCRIPTION
**Pull request by deusebio**
In this PR, we remove the End-2-End tests from the current repository, and moved them to a Test repository, [here](https://github.com/deusebio/repo-test). The reason for this was:

1. Actions are inherently multi-repository, we should have a test repository where we can easily create pre-conditions (e.g. even with tags or branches, without the need to create the files), do commits, merge PRs on which we run actions from a given branch (where the process is implemented). Right now, we should keep on switching branches
2. End-to-end tests interact heavily with the repository (by creating branches, tags, PR), and this would clutter the repository content. Beside, to prevent cluttering, one needs to be able to push force, delete, or other more dev features, that we don't want to enable for a clean, structured and maintanable source repository. Test repositories can have somewhat lighter rules when it comes to the rules of the repositories (e.g. force pushing branches, avoiding signed commits, rolling back). These are all sensible for the base project, but are a bit too restrictive for the test repositories, where we need to play with branches, PRs, etc
3. We should avoid to clutter the main repos with tons of PRs & Triggered Workflows on the main repo as well as Discourse with cluttered content.

The idea discussed with @jdkandersson was to export the E2E tests into a separate repository. Of course this exposes us to the risk of merging features that are not working (altough pass unit and integration tests). The idea here would be to have edge and stable branches in this repository. Scheduled E2E tests workflow run against an edge branch in this repository, whereas actions in production should use stable branches. Promoting edge to stable should be a manual process done once we are comfortable a certain revision is good enough.  

## Tasks

I have now migrated most of the relevant code and business logic about end-to-end testing in the test repository, and I have also extended them to include raising PR and a conflict workflow (only hard conflict so far since the soft ones may go deprecated). This refactoring has however had impacts on this repository as 
1. some small fixes showed up during implementation of E2E tests 
2. we have structured the outputs of the process better (that are used in the E2E tests)
3. Action inputs needed to be extended

There is a [PR](https://github.com/deusebio/repo-test/pull/125/files) which shows some meaningful diff, on how that was achieved, alongside the tests workflow (against the branch of this PR), [here](https://github.com/deusebio/repo-test/actions/runs/5872408897) and [here](https://github.com/deusebio/repo-test/actions/runs/5872408097). The code in the PR can have several improvements, but I believe it would be more critical to get the end-to-end tests up and running.

### CODE REVIEW

**Comment by jdkandersson** [comment]
Should we make the test repository be in the canonical org?

File: action.yaml [hunk 1]
```diff
@@ -42,15 +42,31 @@ inputs:
       the current documentation
     required: true
     type: string
+  base_branch:
+    description: |
+      The name of the base branch for the repository
+    required: false
```
Comment (jdkandersson) [inline action.yaml:48]: Should we include a note here on what the default is?

File: action.yaml [hunk 2]
```diff
@@ -42,15 +42,31 @@ inputs:
...
+    type: string
+  commit_sha:
+    description: |
+      The SHA of the commit to be used when running the action
```
Comment (jdkandersson) [inline action.yaml:52]: Also for this, should we indicate what the default is?

File: main.py [hunk 3]
```diff
@@ -41,18 +43,23 @@ def _parse_env_vars() -> types_.UserInputs:
...
-        # Use the commit SHA if not running as a pull request
-        commit_sha = os.environ["GITHUB_SHA"]
+    if commit_sha is None:
+        try:
+            commit_sha = event["pull_request"]["head"]["sha"]
+        except KeyError:
+            # Use the commit SHA if not running as a pull request
+            commit_sha = os.environ["GITHUB_SHA"]
+
+    logging.info("Base branch: %s", base_branch)
```
Comment (jdkandersson) [inline main.py:62]: Should the commit sha be logged as well?

File: main.py [hunk 4]
```diff
@@ -182,13 +198,35 @@ def main_reconcile(path: Path, user_inputs: types_.UserInputs) -> dict:
...
+
+
 def main() -> None:
     """Execute the action."""
     logging.basicConfig(level=logging.INFO)
 
     # Read input
     user_inputs = _parse_env_vars()
 
+    assert main_checks(user_inputs=user_inputs)  # pylint: disable=E1120
```
Comment (jdkandersson) [inline main.py:228]: What is the disable for please?

File: src/__init__.py [hunk 5]
```diff
@@ -11,46 +11,54 @@
 from . import sort as sort_module
 from .action import DRY_RUN_NAVLINK_LINK, FAIL_NAVLINK_LINK
 from .clients import Clients
-from .constants import DEFAULT_BRANCH, DOCUMENTATION_FOLDER_NAME, DOCUMENTATION_TAG
+from .constants import DOCUMENTATION_FOLDER_NAME, DOCUMENTATION_TAG  # DEFAULT_BRANCH,
```
Comment (jdkandersson) [inline src/__init__.py:14]: Should this comment be here?

File: src/repository.py [hunk 6]
```diff
@@ -163,7 +163,7 @@ def _commit_file_to_tree_element(commit_file: commit_module.FileAction) -> Input
             raise NotImplementedError(f"unsupported file in commit, {commit_file}")
 
 
-class Client:
+class Client:  # pylint: disable=R0904
```
Comment (jdkandersson) [inline src/repository.py:166]: What is this disable please?

File: tests/unit/test___init__.py [hunk 7]
```diff
@@ -1,6 +1,6 @@
 # Copyright 2023 Canonical Ltd.
 # See LICENSE file for licensing details.
-
+# pylint: disable=C0302
```
Comment (jdkandersson) [inline tests/unit/test___init__.py:3]: What is this disable please? I think pylint let's you use the name which makes it easier to figure out what is being disabled. Also, we have a standard for including a note explaining why the disable is needed please

**Comment by jdkandersson** [comment]
It would be great if we can include some documentation on how to execute the new e2e tests and what the new branching strategy is (e.g., merge to `edge` first, run the e2e tests then merge to main or something like that

File: src/repository.py (hunk 6 above)
Comment (deusebio) [reply src/repository.py:166]: This class is definitely becoming to complex (hence the check disabled). We also have an issue about this [here](https://github.com/canonical/upload-charm-docs/issues/175)

File: tests/unit/test___init__.py (hunk 7 above)
Comment (deusebio) [reply tests/unit/test___init__.py:3]: This is somewhat related to the previous comment, where we have classes/files becoming to complex (hence the check disabled). We also have an issue about this https://github.com/canonical/upload-charm-docs/issues/175

BTW...I have changed from code to names

**Comment by deusebio** [comment]
> Should we make the test repository be in the canonical org?

Yes. Let's address the PR and I will then create a brand new repository (I don't think the history we have there is very relevant)

**Review by jdkandersson — APPROVED** [review]

File: main.py [hunk 8]
```diff
@@ -182,18 +198,44 @@ def main_reconcile(path: Path, user_inputs: types_.UserInputs) -> dict:
     return run_reconcile(clients=clients, user_inputs=user_inputs)
 
 
+@execute_in_tmpdir
+def main_checks(path: Path, user_inputs: types_.UserInputs) -> bool:
+    """Checks to make sure that the repository is in a consistent state.
+
```
Comment (yanksyoon) [inline main.py:204]: I'm wondering if this explanation might help :)
```suggestion
    """Checks to make sure that the repository is in a consistent state.
    
      The repository is in a consistent state if there is a `upload-charm-docs/base-content` tag exists in the `base_branch` and the commit belongs to the `base_branch`. If no tag exists, the `upload-charm-docs/base-content` tag will be created for the current commit.
```

**Review by merkata — DISMISSED** [review]

**Review by yanksyoon — DISMISSED** [review]
LGTM, minor nitpicks, please feel free to resolve the conversations to merge!

**Review by jdkandersson — APPROVED** [review]

**Review by yanksyoon — DISMISSED** [review]

**Review by yanksyoon — APPROVED** [review]
LGTM!

**Review by merkata — APPROVED** [review]

**Pull request merged**